llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp

```diff
@@ -84,6 +86,36 @@ ModulePass *llvm::createSYCLLowerWGLocalMemoryLegacyPass() {
   return new SYCLLowerWGLocalMemoryLegacy();
 }
 
+// In sycl header __sycl_allocateLocalMemory builtin call is wrapped in
+// group_local_memory/group_local_memory_for_overwrite functions, which must be
+// inlined first before each __sycl_allocateLocalMemory call can be lowered to a
+// unique global variable. Inlining them here so that this pass doesn't have
+// implicit dependency on AlwaysInlinerPass.
+static bool inlineGroupLocalMemoryFunc(Module &M) {
+  Function *ALMFunc = M.getFunction(SYCL_ALLOCLOCALMEM_CALL);
+  if (!ALMFunc || ALMFunc->use_empty())
+    return false;
+
+  bool Changed = false;
+  for (auto *U : ALMFunc->users()) {
```

Comment:
we need to use a work list here rather than the simple loop.

This function https://github.com/intel/llvm/blob/sycl/sycl/include/syclcompat/memory.hpp#L71 needs to be updated as well, and this function won't be able to handle the nesting. The CI is currently green because there is no test requesting 2 distinct local memory objects using this function in the same kernel.

Reply:
done, thank you for the suggestion. Now I understand what you mean by `syclcompat::local_mem`.
Also added a new test `sycl/test/check_device_code/syclcompat_local_mem.cpp` that has two calls to `syclcompat::local_mem` in a kernel.